buildtest/executors/base.py

```diff
@@ -45,15 +48,97 @@ def __repr__(self):
 
     def get(self, name):
         """Given the name of an executor return the executor for running 
-           a buildtest build.
+           a buildtest build, or get the default.
+        """
+        return self.executors.get(name) or self.executors.get("default")
+
+    def _choose_executor(self, builder):
+        """Choose executor is called at the onset of a run or dryrun. We
+           look at the builder metadata to determine if a default
+           is set for the executor, and fall back to the default.
+
+           :param builder: the builder with the loaded test configuration.
+           :type builder: buildtest.buildsystem.BuilderBase (or subclass).
+        """
+        executor = builder.metadata.get("executor", "default")
+
+        # The executor (or a default) must be define
+        if executor not in self.executors:
+            builder.logger.warning(
+                "executor %s is not defined in default.yml" % executor
+            )
+            sys.exit(1)
+
+        # Get the executor by name, and add the builder to it
+        executor = self.executors.get(executor)
+        executor.builder = builder
+        return executor
+
+    def dry_run(self, builder):
+        """A dry run typically includes all of the steps up to run
+
+           :param builder: the builder with the loaded test configuration.
+           :type builder: buildtest.buildsystem.BuilderBase (or subclass).
+        """
+        # Choose the executor based on the builder provided
+        executor = self._choose_executor(builder)
+
+        # Run each step defined for dry run
+        for step in executor.dryrun_steps:
+            if getattr(executor, step, None):
+                getattr(executor, step)()
+        return executor.result
+
+    def run(self, builder):
+        """Given a buildtest.buildsystem.BuildConfig (subclass) go through the
+           steps defined for the executor to run the build. This should
+           be instantiated by the subclass. For a simple script run, we expect a 
+           setup, build, and finish.
+
+           :param builder: the builder with the loaded test configuration.
+           :type builder: buildtest.buildsystem.BuilderBase (or subclass).
+        """
+        executor = self._choose_executor(builder)
+
+        # Run each step defined for dry run
+        for step in executor.steps:
+            if getattr(executor, step, None):
+                getattr(executor, step)()
+        return executor.result
+
+    def set_default(self, name=None):
```

Comment:
i am a bit confused at how ``set_default`` method is working to get default executor. It's not clear to me yet and would like to understand this with example yaml configuration. 

Right now if i recall we don't have a key ``default`` in the schema under ``executors``. We assume we get the first entry in the list or am i wrong. Looking at your branch configuration https://github.com/researchapps/buildtest-framework/blob/add/executor-steps/buildtest/settings/default.schema.json i don't see any changes made so i take it there is no default executor.

Reply:
Okay, so I'll walk through this. The function there shows the following:

1. If the user has an executor named "default" in their config, this is treated as the default. There is no magic there, it can be of any type.
2. If the user has _no_ executors defined, then we instantiate a BaseExecutor, which is basically a LocalExecutor (take a look at LocalExecutor,it just adds type "local". This could probably easily be LocalExecutor too if you think that is more clear.
3. If the user has at least one executor, the first is taken as the default.

However that flow is agnostic to the command line, and any executor settings in the file! If (at runtime) the user specifies the `--executor` this overrides the default choices above, e.g., this part:

```python
        parser_build.add_argument(
            "-e",
            "--executor",
            help='specify a named executor to use (defaults to "default" key, first in list, then local (base) executor',
        )
```
The argument goes into the instantiation of the BuildExecutor

```
    # Load BuildExecutors
    executor = BuildExecutor(config_opts, default=args.executor)
```
and then that key goes into the same function above:

```
        self.set_default(default)
```
And the same logic persists - the default is set to be that key (or the other options if it doesn't exist).

I set it up exactly like this so that the default executor for most users will bet the first in their list of defined, which (based on our starting config) should be local.

Reply:
Ah, one more detail I forgot (and I know this is important to you) when the BuildExecutor (holding all the executor subtypes) is created, it further chooses the executor based on a setting in the test config file:

```
    def _choose_executor(self, builder):
        """Choose executor is called at the onset of a run or dryrun. We
           look at the builder metadata to determine if a default
           is set for the executor, and fall back to the default.

           :param builder: the builder with the loaded test configuration.
           :type builder: buildtest.buildsystem.BuilderBase (or subclass).
        """
        executor = builder.metadata.get("executor", "default")

        # The executor (or a default) must be define
        if executor not in self.executors:
            builder.logger.warning(
                "executor %s is not defined in default.yml" % executor
            )
            sys.exit(1)

        # Get the executor by name, and add the builder to it
        executor = self.executors.get(executor)
        executor.builder = builder
        return executor
```
I think instead of hard coding "default" I should use the same variable that was possibly set on the command line with `--executor`.

Reply:
Actually scratch that, the "default" executor key is already set to be whatever they specified.

Reply:
Thank you for the detailed explanation, i recall from #251 we decided to have a key ``default`` in executor section (Option 1) so i take it that would be an option key for user but required for schema. 

When you say 

> If the user has an executor named "default" in their config, this is treated as the default. There is no magic there, it can be of any type.

Does that mean we have something like 

```
editor: vi
executors:
  default: runlocal
  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  runlocal:
    type: local
    description: submit jobs locally
```

Or is it we have a default executor something like 

```
editor: vi
executors:
  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  default:
    type: local
    description: submit jobs locally
```

Or something else?

Reply:
I can see value of having ``--executor`` option to run a test with any executor, practically i dont know if test will be written in way to handle that. It out of scope for this discussion but time will tell.

Reply:
@shahzebsiddiqui during the implementation I realized it would get very hard to validate a "special" key value pair for default in the list, and it made more sense to have default just as a key. If you'd like it otherwise please feel free to figure out how to write the schema validator to allow it.

The second example is what I had in mind!

I think command line should generally override all things - it's the highest level of specifying how you want something to run, so I'm not sure I agree with you. I could easily have an entire site worth of test configs that I want to test (with my specific executor) and then I would have to manually change the files to make it work, instead of just using `--executor`. For your example case, the user would be best advised to use different subfolders.

```
buildtest build -c examples/my_executor -e admin
buildtest build -c examples/default_executor
```
Honestly I think we're in a situation where no single solution is going to be perfect for everyone. But generally, when you have config files and command line arguments, the command line args are at the highest level and generally override. It would be very confusing to set a command line `--executor` and have it ignored.

Reply:
> @shahzebsiddiqui during the implementation I realized it would get very hard to validate a "special" key value pair for default in the list

Makes sense to me and would be fine with ``default`` as name of an executor. 

Yea i agree with you the command line supersedes all things and that's what i have observed.